### ISSUE TITLE
Add automatic variable for the pipeline index

### ISSUE THREAD

**Issue by iRon7** [PowerShell/PowerShell#13772]
## Summary of the new feature/enhancement
Initiate and increase an automatic variable (e.g. `$PSIndex`) each time the current object (`$_`/`$Item`) is assigned to another object.

The thought originated from request #13769; *what if there isn't a parameter available to support my index selection requirement?*
E.g.: in case I want to get each odd row to reformat it for a readability (or a combination with #13769 or anything else)
I would need to create a custom index. That index needs to be initiated and increased with each iteration, something like:
```PowerShell
$Index = 0
'A'..'E' | Where-Object { $Index++ -notin 1, 3 }
```
This is actually quiet often required in a lot of situations, as shown in StackOverflow questions along with:
* [Combine single arrays into columns](https://stackoverflow.com/q/23411202/1701026)
* [Make array elements as the keys of hashtable](https://stackoverflow.com/q/60538325/1701026)

## Proposed technical implementation details (optional)
If there would be an automated index variable, I could simplify the solutions to:

### #13769 NotIndex
```PowerShell
'A'..'E' | Where-Object { $PSIndex -notin 1, 3 }
```

### Select odd rows
```PowerShell
'A'..'E' | Where-Object { $PSIndex % 2 }
```

### [Combine single arrays into columns](https://stackoverflow.com/q/23411202/1701026)
```PowerShell
$Column1 = 1..3
$Column2 = 4..6

$Column1 | Foreach-Object {
    [pscustomobject]@{ Column1 = $_; Column2 = $Column2[$PSIndex] }
}
```

### [Make array elements as the keys of hashtable](https://stackoverflow.com/q/60538325/1701026)
```PowerShell
$Keys = 1,2,3
$Values = 'One', 'Two', 'Three'

$HashTable = @{}
$Keys | Foreach-Object {
    [pscustomobject]@{ $HashTable[$_] = $Values[$PSIndex] }
}
```
## Caveats
* I suspect that there already exists a index to support parameters like `-Index` and `-Skip` in the `Select-Object` cmdlet but I have no clue if that is available for every cmdlet in the pipeline
* The indexes at for every cmdlet in the pipeline (and functions like `.Where{ $PSIndex ...}` ?) should all refer to a different index: ` 1..5 | % { $PSIndex } | ... | { $PSIndex }` (as also the case with `$_`/`$PSItem`)

**Comment by IBIT-ZEE** [comment]
it would simplify this too:

https://stackoverflow.com/questions/64481180/intrinsic-index-of-item-in-a-powershell-filtered-item-using-where-object-for-ex

**Comment by mklement0** [comment]
@IBIT-ZEE, no, your use case wouldn't be solved by this, because the proposed `$PSIndex` variable would only be available inside the per-item-processing script blocks.

What you're looking for is a _post_-command way to query all the indices of the objects that `Where-Object` matched.
This probably shouldn't be _automatic_, because the overhead of maintaining this information for a large input collection with many matches can be nontrivial; an _opt-in_ such as with `-MatchingIndexVariable <variableName>` (clunky, but just to illustrate the idea) may be a better approach. I encourage you to open a _new_ feature request.

**Comment by iRon7** [comment]
For this StackOverflow [**In PowerShell is there a way to return the index of an array with substring of a string**](https://stackoverflow.com/q/65029106/1701026) question,
the answer would be something like:

```PowerShell
$myarray | Foreach-Object { if ($_ -match $substring) { $PSIndex } }
```

**Comment by iRon7** [comment]
As suggested in the linked issue [`#14724` implement indexer in loop](https://github.com/PowerShell/PowerShell/issues/14724), I think this purpose should be limited to cmdlets that initiate a new `$_` / `$PSItem` where **`$Collection[$PSIndex] <=> $PSItem`** (as in the original purpose), meaning that `$PSIndex` always falls into step with `$PSItem` (no `$PSItem`, means no `$PSIndex`).

**The way pipelines work it's significantly _more_ important to have an index** because it is possible to retrieve an object from a collection by its index but not visa versa: retrieve the index from a collection by the object. Meaning, a pipeline index should prevent scripters from stalling the whole pipeline by using an index iterator instead (or having to create their own index):

```PowerShell
$Collection = ...                               # this should be avoided as it stalls the pipeline, but is required for:
For ($i = 0; $i -lt $Collection.Count; $i++) {  # or something like (0..($Collection.Count - 1)).ForEach
    $Item = $Collection[$i]                     # But if the $Collection is in memory anyways, this isn't an issue
```

To prevent that `PSIndex` falls apart in nested loops. it might have similar common parameter as [`-PipelineVariable`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_commonparameters) for `$_` / `$PSItem`, something like: **`-PipelineIndex`** (alias `-pi`?).

**Comment by mklement0** [comment]
I agree that an automatic index is _more_ helpful in _pipeline script blocks_ compared to the `foreach` (loop) _statement_, but there's no reason to choose one over the other - both are useful.

I also agree that `$PSIndex` should be limited to contexts where `$_` / `$PSItem` is defined, which, in addition to `ForEach-Object` an `Where-Object` comprises:

* delay-bind script blocks 
* calculated properties (see  https://github.com/PowerShell/PowerShell/issues/14724#issuecomment-776764076 for examples of the latter two)
* the `switch` statement 

A new common `-PipelineIndex` parameter wouldn't work with `switch` (short of introducing such an option there too), and think it's generally fine to stick with `$PSIndex`, given that nested pipelines involve script blocks in which you can easily save the outer `$PSIndex` in another variable (e.g.
`'foo', 'bar', 'baz' | 
ForEach-Object { $ndx = $PSIndex; <# nested pipeline that references $ndx  #> }`)

With the _explicitly named_ automatic index proposed in #14724 in the context of the `foreach` _statement_ (e.g.,
`foreach ($name in 'foo', bar', 'baz'; $index) { "name #$(1+$index): $name" }`), the nesting problem wouldn't even arise.

**Comment by iSazonov** [comment]
It would be more PowerShell-ly if it were a property of this $PSItem.
(I would rather follow the concept of ​​PowerShell than open courses to learn an infinite number of special variables.)

**Comment by iRon7** [comment]
@iSazonov
> ... if it were a property of this $PSItem

Meaning, something like `$_.PSObject.Members.Index`?

**Comment by mklement0** [comment]
`$PSItem` is the input object at hand, and I don't think it is appropriate to add an (instance-level ETS) property to it (for performance reasons alone, but primarily because it modifies the object - unless we use magic, which I think is ill-advised in general and also bears the risk of name collisions with type-native members)

I see a separate variable as the only feasible solution.

Yes, proliferation of automatic variables is a general concern, especially given the potential for name collisions, given that there's no separate namespace for PowerShell-defined variables, which would allow something like `$ps:Index` (see the discussion at https://github.com/PowerShell/PowerShell/issues/4216#issuecomment-315493182).

Currently, prefixing variable names with `PS` is the best solution, so `$PSIndex` seems like a good choice (especially to complement `$PSItem`) - and personally I have no doubt that there is indeed a need for such a variable.

**Comment by iSazonov** [comment]
> @iSazonov
> 
> > ... if it were a property of this $PSItem
> 
> Meaning, something like `$_.PSObject.Members.Index`?

I mean `$PSItem.PipelineIndex`.


> Yes, proliferation of automatic variables is a general concern, especially given the potential for name collisions, given that there's no separate namespace for PowerShell-defined variables, which would allow something like `$ps:Index` (see the discussion at [#4216 (comment)](https://github.com/PowerShell/PowerShell/issues/4216#issuecomment-315493182)).

It is probably worth having a tracking issue for this. I do not believe that MSFT team will waste time on creating new namespace, but we could enhance the `Get-Variable` cmdlet with new parameter `-Kind All | Custom | Preference | System ...` and explicitly enumerate "ill-know" :-) PowerShell variables.

**Comment by mklement0** [comment]
> I mean `$PSItem.PipelineIndex`.

For the reasons stated, I don't think adding a property to `$PSItem` is an option.

> It is probably worth having a tracking issue for this
> enhance the Get-Variable cmdlet with new parameter -Kind All | Custom | Preference | System ...

I like the idea in its own right, but it is obviously not a solution to the name-collision problem.
A related discussion was had in https://github.com/PowerShell/PowerShell/issues/3061#issuecomment-300239754, a long time ago.

At least we have the docs currently ([`about_Automatic_Variables`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Automatic_Variables) and [`about_Preference_Variables`](https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_Preference_Variables))

**Comment by Jaykul** [comment]
For the record: Nothing counts everything that goes through the pipeline.  The `-Index` parameter on `Select-Object` is implemented by [a counter in Select-Object](https://github.com/PowerShell/PowerShell/blob/75b5813f8472b931a0ebcadbf5fec69e1adb3b8f/src/Microsoft.PowerShell.Commands.Utility/commands/utility/Select-Object.cs#L744) which only gets instantiated and incremented when you specify the `-Index` (or `-SkipIndex`) parameter. There can't be a single pipeline counter: the count is different for every command in the pipeline, since any command in the pipeline can output additional things, pass input through, filter input, etc.

Setting that aside: we definitely should not add a counter property to the `$_` pipeline object, because you will step on something. You don't know what properties the thing has, and there's no property you can add which you can guarantee won't interfere with an existing property. For instance, what would happen if you did:

```PowerShell
1..5 | foreach { $_.Index; $_ } | foreach { $_.Index; $_ } 
```

Do you want `$PSIndex` to be set in the `-Begin` block or the `-End` block? I assume not. In that case, I can imagine implementing this as a $PSIndex iteration counter, something that would exist everywhere there's a `$PSItem` ... 

But doing so could be expensive if **every** `foreach` created a counter and incremented it in every loop, even when it's not used. It could even cause overflows if more than `[int]::MaxValue` or `[int64]::MaxValue` items went through the pipeline...

You know that `ForEach-Object` supports multiple process blocks, right:

```PowerShell
'a'..'f' | ForEach-Object @(
    { $i = $j = $k = 0 } # begin
    { $i++; "I:$i = $_"; }
    { $j++; "J:$j = $_" }
    { $k++; "K:$k = $_" }
    { "Total: " + ($i + $j + $k) } # end
) | ForEach-Object @(
    { $l = 0; $m = 0 } # begin
    {
        $l++; "L:$l = $_"; 
        'm'..'p' | ForEach-Object {
             $m++; if($m -gt 75) { "We got $m of these" }
        }
    }
)
```

**Comment by mklement0** [comment]
[The comment above](https://github.com/PowerShell/PowerShell/issues/13772#issuecomment-776869321) summarizes where I think `$PSIndex` should be defined and, yes, it should be tied to where `$_` / `$PSItem` is available.

From that follows the answer to this question:

> Do you want `$PSIndex` to be set in the -Begin block or the -End block? I assume not.

Given that `$_` isn't available in `ForEach-Object` `-Begin` / `-End` blocks, the answer would be no.

However, while definitely pointless in `-Begin` blocks, it _would_ be useful in `-End` blocks, as the count of input objects processed could be inferred from it, analogous to `awk`'s `NR` (1-based row index) variable: `1..3 | awk 'END { print NR }'` -> `3`.

> It could even cause overflows if more than [int]::MaxValue or [int64]::MaxValue

I think we can cross that bridge when we come to it, especially with `[int64]::MaxValue` (roughly `1e18` == a quintillion).

As for performance concerns: I haven't run tests, but I would expect maintaining an index is negligible compared to the cost of invoking PowerShell code (script blocks) iteratively in the pipeline.

If it is a concern nonetheless, perhaps (I can't speak to actual feasibility) references to `$PSIndex` can be detected in the script block(s), from which the (non-)need to maintain `$PSIndex` could be inferred.

**Comment by iRon7** [comment]
@mklement0, thanks for your comments.

As for the *infinite, long-running producer*, if `[int64]::MaxValue`, I guess we are quiet safe for the upcoming thousands of years. The shortest pipeline where I can think about is `... |% {}`, this takes about 3.5 seconds for a million iterations on my PC, meaning it would take me around a million years to reach that on my PC:
```PowerShell
(Measure-Command { 0..999999 |% {} }).TotalSeconds
3.5419796

[int64]::MaxValue / 1000000 * 3.54 / 60 / 60 / 24 / 365 # / sec / min / hours /days 
1035348.07871848 # Years
```
Anyways, if the limit is truly hit, I would simply stop counting which might be less confusing.

@Jaykul,
As with [implement indexer in loop `#14724`](https://github.com/PowerShell/PowerShell/issues/14724), it is *not only* matter of readability, but also protecting scripters from possible pitfalls especially because PowerShell isn't a strict language by default. Knowing that omitting initializers (like `{ $i = $j = $k = 0 } # begin`), might appear to work fine for a novice scripter, but could cause a terrible issue on the long run when the same variable is used in a higher scope (where counters used to get a simply common name like:  `$i` and `$Count`). In other words, a self-initializing index could prevent a PowerShell gotcha.

**Comment by SeeminglyScience** [comment]
> As for performance concerns: I haven't run tests, but I would expect the cost of maintaining an index to be negligible compared to the cost of invoking PowerShell code (script blocks) iteratively in the pipeline.

Maintaining a counter isn't necessarily the concern, assigning a variable can be somewhat expensive on it's own. I can think of a way that it *might* be able to be implemented with minimal impact (subclass `PSVariable` basically) but it'd require a lot of testing.

> If it is a concern nonetheless, perhaps (I can't speak to actual feasibility) references to `$PSIndex` can be detected in the script block(s) on parsing, from which the (non-)need to maintain `$PSIndex` could be inferred.

Static analysis is pretty flimsy in PowerShell.  The variable would probably seem unreliable to a typical PowerShell user (`iex`, `Get-Variable`, etc would all fail to trigger it).

---

Personally I think the use cases for this are too rare enough to warrant the (even if minimal) perf/break risk and work required. Especially when existing solutions aren't terrible.

**Comment by mklement0** [comment]
Thank you for the technical insights, @SeeminglyScience, as usual.
Perhaps needless to say at this point, but personally I think it _is_ worth it; ditto for the related `foreach` proposal, (#14724).

**Comment by GabrielBoehme13** [comment]
> […] existing solutions aren't terrible.

I'd argue that the commonly-suggested solutions I found when looking into this are pretty terrible. I've since created a much better solution (at least for my own needs – more on that later), but I'd argue that PowerShell ought to have something better than "not terrible" built in to address such a common use case. 

I will say, though, that I *don't* think the correct solution is yet another automatic variable – PowerShell has, uh, [*plenty*](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_automatic_variables?view=powershell-7.3) of those already. 😅

I found my way here after trying to manage index variables in my own code related to multiple `ForEach-Object` pipelines. I was following the most commonly-suggested advice (also repeated here), which was to manually manage index variables external to the main script block where the pipeline objects are being handled. Below is a small toy example:

```Powershell
# Example 1
$list = @('first','second','third')
$index = 0 # make sure this variable name doesn't collide with others in this scope!
$list | ForEach-Object {
  "[$index] $_"
  $index += 1 # don't forget to increment! (and be sure it's the correct index variable!)
}

<# output:
[0] first
[1] second
[2] third
#>
```

While this can work in very simple scenarios like this, it's messy and error-prone, and I quickly discovered it that didn't scale well at all. I found myself agreeing with the concerns @iRon7 [mentioned](https://github.com/PowerShell/PowerShell/issues/13772#issuecomment-815564294) – this approach leads us to [the Pit of Despair, not the Pit of Success](https://blog.codinghorror.com/falling-into-the-pit-of-success/).

After reading a [comment](https://github.com/PowerShell/PowerShell/issues/13772#issuecomment-777564569) from @iSazonov about wanting a more "PowerShell-ly" solution, I realized there was a much better way to do this within existing PowerShell functionality – transforming objects in the pipeline into custom objects / hashtables with .Index and .Value properties. Below is my (very rough-and-ready) implementation of this idea:

```Powershell
# Example 2

function indexify { # naming things is hard
  $index = 0 # each call creates and manages its own index variable - no collisions
  $input | ForEach-Object { # not sure this is the "correct" way to do this, but it works!
    @{
      Index = $index
      Value = $_
    }
    $index += 1
  }
}

$list = @('first','second','third')
$list | indexify | ForEach-Object {
  "[$($_.Index)] $($_.Value)"
}

# produces the same output as Example 1
```

I can now add `indexify` to pipelines anywhere in my code, and I get index/value pairs without having to manually name or increment any index variables – Pit of Success unlocked! 🎉 

While I'm happy with the solution I've created (as it seems to address my current needs perfectly), I shouldn't have needed to figure this out on my own. *Some* built-in way for pipelines to opt into automated indexing ought to be available within PowerShell itself, so the most common online advice on how to do this doesn't keep pointing people to the Pit of Despair.

**Comment by wildone** [comment]
will try this on 1mill elements :D

**Comment by beau-witter** [comment]
I got here because I was sure that by now Powershell had a built-in way to keep track of the current index within something like a `ForEach-Object` loop, only to find out that it does not and there are some very interesting thoughts about it from both sides. I really like @GabrielBoehme13 's `indexify` (I agree naming is hard) function, but ultimately just wish there was something supported within PowerShell itself to know current index without having to resort to a `for` loop or maintaining the index manually as shown in Example 1.

**Comment by iRon7** [comment]
If you consider a custom/self made solution, you might create a [proxy command](https://stackoverflow.com/a/77635453/1701026) based on the existing [ForEach-Object](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/foreach-object) cmdlet.

### [For-Object](https://github.com/iRon7/For-Object)

```PowerShell
Install-Script -Name For-Object

'One', 'Two', 'Three' | For-Object { "[$PSIndex] $PSItem" }
[0] One
[1] Two
[2] Three
```

**Comment by iRon7** [comment]
A pretty close alternative to the request is using the `$local:` [scope modifier](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_scopes#scope-modifiers), see: https://github.com/PowerShell/PowerShell/issues/26645#issuecomment-3694738788:

### #13769 NotIndex
```PowerShell
'A'..'E' | Where-Object { $Local:Index++ -notin 1, 3 }
```

### Select odd rows
```PowerShell
'A'..'E' | Where-Object { $Local:Index++ % 2 }
```

### [Combine single arrays into columns](https://stackoverflow.com/q/23411202/1701026)
```PowerShell
$A1= 1..3
$A2= 4..6
$A3= 7..9

$Column1 | Foreach-Object {
    [pscustomobject]@{ A1 = $_; A2= $A2[$Local:i++]; A3 = $A3[$Local:j++] }
}
```

### [Make array elements as the keys of hashtable](https://stackoverflow.com/q/60538325/1701026)
```PowerShell
$Keys = 1,2,3
$Values = 'One', 'Two', 'Three'

$HashTable = @{}
$Keys | Foreach-Object { $HashTable[$_] = $Values[$Local:Index++] }
```

---
#### Update

There is actually a pitfall in this suggestion and using the `$Local:` modifier for "self-initializing" variable (index):

> [!WARNING]
>
> The `$Local:` modifier will only work once as a "self-initializing" variable! This means that it won't reset to its default value when reused (e.g. when embedded in another loop) in the same function or script (scope).

In other words, this:

```PowerShell
1..3 | ForEach-Object {
    'A'..'E' | Where-Object { $Local:Index++ -notin 1, 3 }
}
```

Works only the first iteration as expected... (see also: #26645)
